### PR TITLE
jprofiler: new package

### PR DESCRIPTION
**Pull request by catap**
###### Description of changes

Very popular java profiler which is missed for NixOS.

Homepage: https://www.ej-technologies.com/products/jprofiler/overview.html

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:25]:
```suggestion
  desktopItem = makeDesktopItem {
    name = pname;
    exec = pname;
    icon = pname;
    comment = meta.description;
    desktopName = "JProfiler";
    genericName = "Java Profiler Tool";
    categories = [ "Development" ];
  };
```

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:39]:
```suggestion
    for i in $out/bin/*; then
      wrapProgram $i --set JAVA_HOME "${jdk.home}" 
    fi
```

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:57]: Given that it is written in Java, could this also work for other platforms?

**Review by azahi — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:57]: @azahi it supports different platforms but via different distributives which includes some different pre-build library which is used via JNA/JNI interface.

It is possible to install from linux's `.tar.gz` to macOS for example but some feature might not work.

To be honest I never tried.

Comment (azahi) [reply pkgs/development/tools/java/jprofiler/default.nix:52]: I see. I think you can leave the platform as-is or we can just ask someone to test it out on MacOS before merging.

**Comment by catap** [comment]
@azahi I've integrated all your suggestions and rebased to the last master.

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 178813` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>jprofiler</li>
  </ul>
</details>

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [reply pkgs/development/tools/java/jprofiler/default.nix]: I think this broke the build because it's supposed to be 
```
    for i in $out/bin/*; do
      wrapProgram $i --set JAVA_HOME "${jdk.home}" 
    done
```
My bad.

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:36]:
```suggestion
    for i in $out/bin/*; do
      wrapProgram $i --set JAVA_HOME "${jdk.home}"
    done
```

**Review by azahi — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix]: @azahi and I've missed that. Thanks for fixing it.

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 178813` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>jprofiler</li>
  </ul>
</details>

**Comment by azahi** [comment]
Fails for me with:
```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/6vmrchkxpplp2ds8400awb5kgayf18zj-jprofiler_linux_13_0_2.tar.gz
source root is jprofiler13.0.2
setting SOURCE_DATE_EPOCH to timestamp 1651651540 of file jprofiler13.0.2/release
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
no Makefile, doing nothing
@nix { "action": "setPhase", "phase": "installPhase" }
installing

Builder called die: Cannot wrap '/nix/store/yw0zrz2dvnns45syh66n02bzzagjx6vz-jprofiler-13.0.2/bin/agent.jar' because it is not an executable file
Backtrace:
7 assertExecutable /nix/store/q12qdr07fhpgl0n2zd4js56brbydk1s4-hook/nix-support/setup-hook
207 wrapProgramShell /nix/store/q12qdr07fhpgl0n2zd4js56brbydk1s4-hook/nix-support/setup-hook
202 wrapProgram /nix/store/q12qdr07fhpgl0n2zd4js56brbydk1s4-hook/nix-support/setup-hook
1393 genericBuild /nix/store/n4rm9iyf5prbgxzjhv8gsxmxcxprga9x-stdenv-linux/setup
2 main /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh
```

**Comment by catap** [comment]
@azahi I've wrap now only really executable files. It should fix an issue for now.

**Review by azahi — CHANGES_REQUESTED** [review]
I've also noticed the sheer size of the resulting derivation. There are a lot of IMO unneeded code examples and various binary files, especially ones that target different architectures. We probably shouldn't include pre-compiled libraries for AArch64 when building a derivation for x86_64. Given that the derivation would be used to only execute jprofiler binaries, we probably should just install `bin` directory, omitting `api`, `demo` and etc.  

The bundled launch scripts also appear to be broken for me:
```
No suitable Java Virtual Machine could be found on your system.
The version of the JVM must be at least 11 and at most 17.
Please define INSTALL4J_JAVA_HOME to point to a suitable JVM.
```
Looking at the mess of what are these bundled scripts, the `INSTALL4J_JAVA_HOME_OVERRIDE` could be specified at the top to pass the path to `${jdk.home}`. You could also ditch `wrapProgram` altogether and work with these scripts directly.

You've picked a very complicated package as your first contribution to Nixpkgs :(

**Comment by catap** [comment]
@azahi I've added support of macOS. Unfortently I can't use the last version because `undmg` can't open it. But one day.. maybe..

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:7]: If you need to specify a certain JDK version, you should do that in `all-packages.nix`...
```suggestion
, jdk };
```

File: pkgs/top-level/all-packages.nix
Comment (azahi) [inline pkgs/top-level/all-packages.nix:16673]: ...Like this:
```suggestion
  jprofiler = callPackage ../development/tools/java/jprofiler {
    jdk = jdk11;
  };
```

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:58]: Please add `runHook`s.

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:59]: You shouldn't do that. As I've mentioned earlier, this install _**a lot**_ of unnecessary files. If you think that we should keep docs or examples, do this in a separate derivation [output](https://nixos.org/manual/nixpkgs/stable/#sec-multiple-outputs-). Also make sure that there are no redundant libraries for other architectures/platforms.

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:15]: I haven't seen platform-dependent versions anywhere in Nixpkgs, especially major bumps. @SuperSandro2000 can we allow this?

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [inline pkgs/development/tools/java/jprofiler/default.nix:85]: Please don't leave commented out code, it will only confuse people. If you think that this must be kept, add a comment explaining why.
```suggestion
```

**Review by azahi — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:15]: @azahi see example here: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/libraries/science/math/mkl/default.nix

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:59]: The only things that I can remove is a few libraries for unsupported by nixos architectures. It saves something like 10%-15% of diskspace — nothing from my point of view.

"samples" are available via UI
Screenshot 2022-07-01 at 01 24 34
 of JProfiler => I really don't think that it is a good idea to remove them:
Screenshot 2022-07-01 at 01 24 39

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [reply pkgs/development/tools/java/jprofiler/default.nix:15]: OK, I guess it's alright then.

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (azahi) [reply pkgs/development/tools/java/jprofiler/default.nix:60]: It's not the issue of occupied space (although, this could be annoying given that Hydra could rebuild this package frequently and waste a lot of unneeded space across multiple derivations). The issue here is the files that are not supposed to be there because they are functionally meaningless, as the case with libraries and executable which will never run. 

As for examples and documentations, I still think that a separate output would be preferable. Your justification to use examples from the GUI makes sense too. I'll leave it up to you.

**Comment by azahi** [comment]
Result of `nixpkgs-review pr 178813` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>jprofiler</li>
  </ul>
</details>

**Comment by azahi** [comment]
Works fine I guess, but I'm still not happy about the resulting output structure. I guess we could invite more people to review this. Maybe I'm just being nitpicky.

**Comment by azahi** [comment]
I also forgot to mention that your commit message is wrong. It should be `jprofile: init at <version>`.

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix]: @azahi I've tested it on macOS ;) I'm running NixOS inside Parallels and using macOS as host system.

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (catap) [reply pkgs/development/tools/java/jprofiler/default.nix:60]: @azahi see example like intellij-idea which. It is quite standalone tool.

**Comment by SuperSandro2000** [comment]
Please name your commits according to the contributing guide

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:30]: Please make the entire fetcher conditionnal, not only the arguments to fetchurl.

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:31]: Please use lib.replaceStrings

File: pkgs/development/tools/java/jprofiler/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/java/jprofiler/default.nix:72]: Use installDesktopItems

**Comment by catap** [comment]
@SuperSandro2000 it's done

**Pull request merged**
